File: 2022/OpenCV4 tutorial/17 Threshold Segmentation/Range.py
import cv2 as cv
import numpy as np
def simple_threshold():
    src = cv.imread("merge_split.jpg")
    dst = cv.cvtColor(src, cv.COLOR_BGR2HSV)
    mask = cv.inRange(dst, np.array([28, 100, 100]), np.array([40, 255, 255]))
    cv.imshow("mask", mask)

    for i in range(0, 3):
        dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)

    dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
    cv.imshow("yellow", dst)
    cv.waitKey()
def main():
    src = cv.imread("merge_split.jpg")
    hsv = cv.cvtColor(src, cv.COLOR_BGR2HSV)
    para4Th = {'hL': 0, 'hU': 255, 'sL': 0, 'sU': 255, 'vL': 0, 'vU': 255}
    cv.namedWindow("inRange")
    def hLow_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['hL'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)

    def hUp_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['hU'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)

    def sLow_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['sL'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)

    def sUp_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['sU'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)

    def vLow_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['vL'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)

    def vUp_onChange(pos):
        # 克隆图像
        dst = np.zeros(hsv.shape, dtype=np.uint8)
        np.copyto(dst, hsv)

        para4Th['vU'] = pos
        mask = cv.inRange(dst, np.array([para4Th['hL'], para4Th['sL'], para4Th['vL']]), np.array([para4Th['hU'], para4Th['sU'], para4Th['vU']]))
        for i in range(0, 3):
            dst[:, :, i] = cv.bitwise_and(src[:, :, i], mask)
        dst = cv.addWeighted(src, 0.2, dst, 0.8, 0)
        cv.imshow("inRange", dst)
    cv.createTrackbar("hL", "inRange", 0, 255, hLow_onChange)  # 创建滑杆
    cv.createTrackbar("sL", "inRange", 0, 255, sLow_onChange)  # 创建滑杆
    cv.createTrackbar("vL", "inRange", 0, 255, vLow_onChange)  # 创建滑杆
    cv.createTrackbar("hU", "inRange", 255, 255, hUp_onChange)  # 创建滑杆
    cv.createTrackbar("sU", "inRange", 255, 255, sUp_onChange)  # 创建滑杆
    cv.createTrackbar("vU", "inRange", 255, 255, vUp_onChange)  # 创建滑杆



    cv.waitKey()

if __name__ == '__main__':
    main()

File: 2022/OpenCV4 tutorial/17 Threshold Segmentation/Range.cs
using System;
using System.IO;
using OpenCvSharp; //导入OpenCV4
using System.Collections.Generic;

namespace Gradient
{

    class Program
    {
        static void simple_test()
        {
            Mat inputImage = Cv2.ImRead("merge_split.jpg"),hsv = new Mat() ,dst = new Mat(), mask = new Mat();
            if (!inputImage.Empty())
            {
                Cv2.CvtColor(inputImage, hsv, ColorConversionCodes.BGR2HSV);
                Cv2.InRange(hsv, new Scalar(28, 100, 100), new Scalar(40, 255, 255), mask);
                Cv2.ImShow("mask", mask);

                Mat[] bgr = new Mat[3], bgr2 = new Mat[3];
                for (int i = 0; i < 3; i++)
                {
                    bgr2[i] = Mat.Zeros(MatType.CV_8UC1, new int[] { inputImage.Cols, inputImage.Rows });
                }
                bgr = Cv2.Split(inputImage);

                for (int i = 0; i < 3; i++)
                {
                    Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                }
                Cv2.Merge(bgr2, dst);
                Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                Cv2.ImShow("inRange", dst);
                Cv2.WaitKey();
            }

        }
        static void Main(string[] args)
        {
            DirectoryInfo rootDir = Directory.GetParent(Environment.CurrentDirectory);
            string root = rootDir.Parent.Parent.FullName;
            System.IO.Directory.SetCurrentDirectory(root);

            Mat inputImage = Cv2.ImRead("merge_split.jpg");
            if (!inputImage.Empty())
            {
                Mat hsv = new Mat(), dst = new Mat(), mask = new Mat();
                Cv2.CvtColor(inputImage, hsv, ColorConversionCodes.BGR2HSV);
                Dictionary<string, int> range_data = new Dictionary<string, int>
                {
                    { "hL", 0 },
                    { "sL", 0 },
                    { "vL", 0 },
                    { "hU", 255 },
                    { "sU", 255 },
                    { "vU", 255 }
                };
                Cv2.NamedWindow("inRange", WindowFlags.AutoSize);
                Mat[] bgr = new Mat[3], bgr2 = new Mat[3];
                for (int i = 0; i < 3; i++)
                {
                    bgr2[i] = Mat.Zeros(MatType.CV_8UC1, new int[] { inputImage.Cols, inputImage.Rows });
                }
                bgr = Cv2.Split(inputImage);
                Cv2.CreateTrackbar("hLow", "inRange", 255, (int hL, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["hL"] = hL;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );

                Cv2.CreateTrackbar("sLow", "inRange", 255, (int sL, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["sL"] = sL;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );
                Cv2.CreateTrackbar("vLow", "inRange", 255, (int vL, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["vL"] = vL;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );
                Cv2.CreateTrackbar("hUp", "inRange", 255, (int hU, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["hU"] = hU;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );

                Cv2.CreateTrackbar("sUp", "inRange", 255, (int sU, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["sU"] = sU;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );
                Cv2.CreateTrackbar("vUp", "inRange", 255, (int vU, IntPtr userData) =>
                {
                    Mat dst = hsv.Clone();
                    range_data["vU"] = vU;
                    Cv2.InRange(hsv, new Scalar(range_data["hL"], range_data["sL"], range_data["vL"]), new Scalar(range_data["hU"], range_data["sU"], range_data["vU"]), mask);

                    for (int i = 0; i < 3; i++)
                    {
                        Cv2.BitwiseAnd(bgr[i], mask, bgr2[i]);
                    }
                    Cv2.Merge(bgr2, dst);
                    Cv2.AddWeighted(inputImage, 0.2, dst, 0.8, 0, dst);
                    Cv2.ImShow("inRange", dst);

                }
                );
                Cv2.SetTrackbarPos("hLow", "inRange", 0);
                Cv2.SetTrackbarPos("sLow", "inRange", 0);
                Cv2.SetTrackbarPos("vLow", "inRange", 0);
                Cv2.SetTrackbarPos("hUp", "inRange", 255);
                Cv2.SetTrackbarPos("sUp", "inRange", 255);
                Cv2.SetTrackbarPos("vUp", "inRange", 255);
                Cv2.WaitKey();

            }
        }
    }
}